x/tokenomics/keeper/token_logic_modules.go

```diff
@@ -130,109 +150,118 @@ func (k Keeper) ProcessTokenLogicModules(
-	// blocker.
-
-	// Determine the total number of tokens that'll be used for settling the session.
-	// When the network achieves equilibrium, this will be the mint & burn.
-	settlementCoin, err = k.numRelaysToCoin(ctx, numRelays, &service)
+	// Determine the total number of tokens being claimed (i.e. requested)
+	// by the supplier for the amount of work they did to service the application
+	// in the session.
+	claimSettlementCoin, err = k.numRelaysToCoin(ctx, numRelays, &service)
 	if err != nil {
 		return err
 	}
 
-	// Retrieving the relay mining difficulty for the service at hand
-	relayMiningDifficulty, found := k.GetRelayMiningDifficulty(ctx, service.Id)
-	if !found {
-		if err != nil {
-			return err
-		}
-		logger.Warn(fmt.Sprintf("relay mining difficulty for service %q not found. Using default difficulty", service.Id))
-		relayMiningDifficulty = tokenomicstypes.RelayMiningDifficulty{
-			ServiceId:    service.Id,
-			BlockHeight:  sdk.UnwrapSDKContext(ctx).BlockHeight(),
-			NumRelaysEma: numRelays,
-			TargetHash:   prooftypes.DefaultRelayDifficultyTargetHash,
-		}
-	}
-
 	// Helpers for logging the same metadata throughout this function calls
 	logger = logger.With(
 		"num_relays", numRelays,
-		"num_settlement_upokt", settlementCoin.Amount,
+		"claim_settlement_upokt", claimSettlementCoin.Amount,
 		"session_id", sessionHeader.GetSessionId(),
 		"service_id", sessionHeader.GetService().Id,
 		"supplier_operator", supplier.OperatorAddress,
 		"application", application.Address,
 	)
-	logger.Info(fmt.Sprintf("About to start processing TLMs for (%d) relays equaling to (%s) coins", numRelays, settlementCoin))
 
+	// Retrieving the relay mining difficulty for the service at hand
+	relayMiningDifficulty, found := k.GetRelayMiningDifficulty(ctx, service.Id)
+	if !found {
+		relayMiningDifficulty = newDefaultRelayMiningDifficulty(ctx, logger, service.Id, numRelays)
+	}
+
+	// Ensure the claim amount is within the limits set by Relay Mining.
+	// Update the settlement amount if not and emit any necessary events in doing so.
+	actualSettlementCoin, err := k.ensureClaimAmountLimits(ctx, logger, &application, &supplier, claimSettlementCoin)
+	if err != nil {
+		return err
+	}
+	logger = logger.With("actual_settlement_upokt", actualSettlementCoin)
+
+	logger.Info(fmt.Sprintf("About to start processing TLMs for (%d) relays equaling to (%s) upokt claimed", numRelays, actualSettlementCoin))
 	// Execute all the token logic modules processors
 	for tlm, tlmProcessor := range tokenLogicModuleProcessorMap {
-		logger.Info(fmt.Sprintf("Starting to execute TLM %q", tlm))
-		if err := tlmProcessor(k, ctx, &service, &application, &supplier, settlementCoin, &relayMiningDifficulty); err != nil {
+		logger.Info(fmt.Sprintf("Starting TLM processing: %q", tlm))
+		if err := tlmProcessor(k, ctx, &service, claim.SessionHeader, &application, &supplier, actualSettlementCoin, &relayMiningDifficulty); err != nil {
 			return err
```

Comment:
I feel like there should be a sentinel error which wraps the TLM name and error here:

```suggestion
			return ErrTokenomicsTLMError.Wrapf("TLM %q: %v", tlm err)
```